srml/staking/src/lib.rs

```diff
@@ -299,118 +490,98 @@ impl<T: Trait> Module<T> {
+		// The amount we are actually going to slash (can't be bigger than thair total exposure)
+		let slash = slash.min(exposure.total);
+		// The amount we'll slash from the validator's stash directly.
+		let own_slash = exposure.own.min(slash);
+		let own_slash = own_slash - T::Currency::slash(v, own_slash).unwrap_or_default();
+		// The amount remaining that we can't slash from the validator, that must be taken from the nominators.
+		let rest_slash = slash - own_slash;
+
+		if !rest_slash.is_zero() {
+			// The total to be slashed from the nominators.
+			let total = exposure.total - exposure.own;
 			if !total.is_zero() {
-				let safe_mul_rational = |b| b * rem / total;// FIXME #1572 avoid overflow
-				for n in noms.iter() {
-					let _ = T::Currency::slash(n, safe_mul_rational(T::Currency::total_balance(n)));	// best effort - not much that can be done on fail.
+				let safe_mul_rational = |b| b * rest_slash / total;// FIXME #1572 avoid overflow
+				for i in exposure.others.iter() {
+					let _ = T::Currency::slash(&i.who, safe_mul_rational(i.value));	// best effort - not much that can be done on fail.
 				}
 			}
 		}
 	}
 
 	/// Reward a given validator by a specific amount. Add the reward to their, and their nominators'
-	/// balance, pro-rata.
+	/// balance, pro-rata based on their exposure, after having removed the validator's pre-payout cut.
 	fn reward_validator(who: &T::AccountId, reward: BalanceOf<T>) {
-		let off_the_table = reward.min(Self::validator_preferences(who).validator_payment);
+		let off_the_table = reward.min(Self::validators(who).validator_payment);
 		let reward = reward - off_the_table;
 		let validator_cut = if reward.is_zero() {
 			Zero::zero()
 		} else {
-			let noms = Self::current_nominators_for(who);
-			let total = noms.iter()
-				.map(T::Currency::total_balance)
-				.fold(T::Currency::total_balance(who), |acc, x| acc + x)
-				.max(One::one());
+			let exposure = Self::stakers(who);
+			let total = exposure.total.max(One::one());
 			let safe_mul_rational = |b| b * reward / total;// FIXME #1572:  avoid overflow
-			for n in noms.iter() {
-				let _ = T::Currency::reward(n, safe_mul_rational(T::Currency::total_balance(n)));
+			for i in exposure.others.iter() {
+				let amount = safe_mul_rational(i.value);
+				let payee = Self::payee(&i.who);
+				match payee {
+					RewardDestination::Controller => {
+						let _ = T::Currency::reward(&i.who, amount);
+					}
+					RewardDestination::Stash => {
+						let _ = Self::ledger(&i.who).map(|l| T::Currency::reward(&l.stash, amount));
+					}
+				 	RewardDestination::Staked => <Ledger<T>>::mutate(&i.who, |ml| {
+						if let Some(l) = ml.as_mut() {
+							l.active += amount;
+							l.total += amount;
+							let _ = T::Currency::reward(&l.stash, amount);
+						}
+					}),
+				}
 			}
-			safe_mul_rational(T::Currency::total_balance(who))
+			safe_mul_rational(exposure.own)
 		};
 		let _ = T::Currency::reward(who, validator_cut + off_the_table);
```

Comment:
Why doesn't this reward to the payee?

Reply:
`let _ = T::Currency::reward(who, validator_cut + off_the_table);` is the portion of the reward which goes to the validator, while that whole config for `RewardDestination` is about how **nominators** are paid, hence this reward payment that you mention is going directly to the validator account. 

Suggestion: To avoid confusion `RewardDestination` can be called `NominatorRewardDestination`. 

And: Since the nominator does have the freedom to forward its own rewards to its associative stash account (@shawntabrizi note that both a controller entity and a nominator entity is the combination a two account, one stash, and one controller; see `nominating_and_rewards_should_work()`), I would say it is also fair to maybe give the validator the same ability? Unless if we want to stick to the rule: _Validator's stash account is frozen and shall not be mutated_ @gavofyork

Reply:
Yeah, my question is more philosophical, as in, why don't we have the logic support Validators also paying to the RewardDestination?

Reply:
Then I am in on the same page.

Reply:
it's a bug - the same logic should be used for the validator, too.